ze/sampling_daemon.c

```diff
@@ -0,0 +1,803 @@
+    return;
+  ze_result_t result;
+  for (uint32_t engineIdx = 0; engineIdx < _sampling_engineCounts[driverIdx][deviceIdx];
+       ++engineIdx) {
+    zes_engine_stats_t engineStats = {0};
+    result = ZES_ENGINE_GET_ACTIVITY_PTR(_sampling_engineHandles[driverIdx][deviceIdx][engineIdx],
+                                         &engineStats);
+    if (result != ZE_RESULT_SUCCESS) {
+      _ZE_ERROR_MSG("ZES_ENGINE_GET_ACTIVITY_PTR", result);
+      continue;
+    }
+    do_tracepoint(lttng_ust_ze_sampling, engineStats,
+                  (zes_device_handle_t)_sampling_hDevices[driverIdx][deviceIdx],
+                  (zes_engine_handle_t)_sampling_engineHandles[driverIdx][deviceIdx][engineIdx],
+                  &engineStats);
+  }
+}
+
+static void thapi_sampling_energy() {
+  for (uint32_t driverIdx = 0; driverIdx < _sampling_driverCount; driverIdx++) {
+    for (uint32_t deviceIdx = 0; deviceIdx < _sampling_deviceCount[driverIdx]; deviceIdx++) {
+      if (tracepoint_enabled(lttng_ust_ze_sampling, gpu_frequency)) {
+        readFrequency_dump(driverIdx, deviceIdx);
+      }
+      if (tracepoint_enabled(lttng_ust_ze_sampling, gpu_energy)) {
+        readEnergy_dump(driverIdx, deviceIdx);
+      }
+      if (tracepoint_enabled(lttng_ust_ze_sampling, engineStats)) {
+        readEngines_dump(driverIdx, deviceIdx);
+      }
+      if (tracepoint_enabled(lttng_ust_ze_sampling, fabricPort)) {
+        readFabricPorts_dump(driverIdx, deviceIdx);
+      }
+      if (tracepoint_enabled(lttng_ust_ze_sampling, memStats)) {
+        readMemModules_dump(driverIdx, deviceIdx);
+      }
+    }
+  }
+}
+void process_sampling() {
+
+  struct timespec interval;
+  interval.tv_sec = 0;
+  interval.tv_nsec = 50000000; // 50ms interval
+  thapi_sampling_energy();
+  _sampling_handle = thapi_register_sampling(&thapi_sampling_energy, &interval);
+}
+void cleanup_sampling() {
+  if (_sampling_handle) {
+    thapi_unregister_sampling(_sampling_handle);
+    _sampling_handle = NULL;
+  }
+}
+
+void signal_handler(int signum) {
+  if (signum == SIG_SAMPLING_FINISH) {
+    cleanup_sampling();
+    running = false;
+  }
+}
+
+int main(int argc, char **argv) {
+
+  if (argc < 2) {
+    _USAGE_MSG("<parent_pid>", argv[0]);
+    return 1;
+  }
+  int parent_pid = atoi(argv[1]);
+  if (parent_pid <= 0) {
+     _ERROR_MSG("Invalid or missing parent PID. A positive integer is required.");
+     return 1;
+  }
+  int verbose = 0;
+  thapi_sampling_init();// Initialize sampling
+
+  // Load necessary libraries
+  void *handle = NULL;
+  char *s = getenv("LTTNG_UST_ZE_LIBZE_LOADER");
+  if (s) {
+    handle = dlopen(s, RTLD_LAZY | RTLD_LOCAL | RTLD_DEEPBIND);
+  } else {
+    handle = dlopen("libze_loader.so", RTLD_LAZY | RTLD_LOCAL | RTLD_DEEPBIND);
+  }
+
+  if (!handle) {
+    _DL_ERROR_MSG();
+    return 1;
+  }
+  //Find zes symbols
+  find_ze_symbols(handle, verbose);
+  //Initialize device and telemetry handles
+  initializeHandles();
```

Comment:
You need to send `Ready` have `initializeHandles` no?

Reply:
I send READY after initializing 
if (kill(parent_pid, SIG_SAMPLING_READY) != 0) {
    _ERROR_MSG("Failed to send READY signal to parent");
  }